aggregator/entry.go

```diff
@@ -0,0 +1,243 @@
+}
+
+// NewEntry creates a new entry
+func NewEntry(lists *MetricLists, opts Options) *Entry {
+	e := &Entry{
+		opts:         opts,
+		aggregations: make(map[policy.Policy]*list.Element),
+	}
+	e.ResetSetData(lists)
+	return e
+}
+
+// IncWriter increases the writer count
+func (e *Entry) IncWriter() { atomic.AddInt32(&e.numWriters, 1) }
+
+// DecWriter decreases the writer count
+func (e *Entry) DecWriter() { atomic.AddInt32(&e.numWriters, -1) }
+
+// ResetSetData resets the entry and sets initial data
+func (e *Entry) ResetSetData(lists *MetricLists) {
+	e.closed = false
+	e.lists = lists
+	e.version = policy.InitPolicyVersion
+	e.numWriters = 0
+	e.recordLastAccessed(e.opts.ClockOptions().NowFn()())
+}
+
+// AddMetricWithPolicies adds a metric along with applicable policies
+func (e *Entry) AddMetricWithPolicies(
+	mu unaggregated.MetricUnion,
+	policies policy.VersionedPolicies,
+) error {
+	// NB(xichen): it is important that we determine the current time
+	// within the time lock. This ensures time ordering by wrapping
+	// actions that need to happen before a given time within a read lock,
+	// so it is guaranteed that actions before when a write lock is acquired
+	// must have all completed. This is used to ensure we never write metrics
+	// for times that have already been flushed.
+	timeLock := e.opts.TimeLock()
+	timeLock.RLock()
+	defer timeLock.RUnlock()
+
+	e.RLock()
+	if e.closed {
+		e.RUnlock()
+		return errEntryClosed
+	}
+
+	currTime := e.opts.ClockOptions().NowFn()()
+	e.recordLastAccessed(currTime)
+
+	if !e.shouldUpdatePoliciesWithLock(currTime, policies.Version, policies.Cutover) {
+		e.addMetricWithLock(currTime, mu)
+		e.RUnlock()
+		return nil
+	}
+	e.RUnlock()
+
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return errEntryClosed
+	}
+
+	if e.shouldUpdatePoliciesWithLock(currTime, policies.Version, policies.Cutover) {
+		if err := e.updatePoliciesWithLock(mu.Type, mu.ID, policies.Version, policies.Policies); err != nil {
+			// NB(xichen): if an error occurred during policy update, the policies
+			// will remain as they are, i.e., there are no half-updated policies
+			e.Unlock()
+			return err
+		}
+	}
+
+	e.addMetricWithLock(currTime, mu)
+	e.Unlock()
+
+	return nil
+}
+
+// ShouldExpire returns whether the entry should expire
+func (e *Entry) ShouldExpire(now time.Time) bool {
+	e.RLock()
+	if e.closed {
+		e.RUnlock()
+		return false
+	}
+	// Only expire the entry if there are no active writers
+	// and it has reached its ttl since last accessed
+	shouldExpire := e.writerCount() == 0 && now.After(e.lastAccessed().Add(e.opts.EntryTTL()))
+	e.RUnlock()
+	return shouldExpire
+}
+
+// Expire expires the entry
+func (e *Entry) Expire() {
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return
+	}
+	e.closed = true
+	for p, agg := range e.aggregations {
+		switch elem := agg.Value.(type) {
+		case metricElem:
+			elem.MarkAsTombstoned()
+		default:
+			// We should never get here
```

Comment:
Fair enough, was trying to make the error message a bit nicer but yeah can get rid of it